streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala

```diff
@@ -0,0 +1,203 @@
+    val kvMapper = mapper.tupled andThen tuple2ToKeyValue
+    inner.map[KR, VR]((k, v) => kvMapper(k,v))
+  }
+
+  def mapValues[VR](mapper: V => VR): KStream[K, VR] = {
+    inner.mapValues[VR](mapper.asValueMapper)
+  }
+
+  def mapValues[VR](mapper: (K, V) => VR): KStream[K, VR] = {
+    inner.mapValues[VR](mapper.asValueMapperWithKey)
+  }
+
+  def flatMap[KR, VR](mapper: (K, V) => Iterable[(KR, VR)]): KStream[KR, VR] = {
+    val kvMapper = mapper.tupled andThen (iter => iter.map(tuple2ToKeyValue).asJava)
+    inner.flatMap[KR, VR]((k,v) => kvMapper(k , v))
+  }
+
+  def flatMapValues[VR](processor: V => Iterable[VR]): KStream[K, VR] = {
+    inner.flatMapValues[VR](processor.asValueMapper)
+  }
+
+  def flatMapValues[VR](processor: (K, V) => Iterable[VR]): KStream[K, VR] = {
+    inner.flatMapValues[VR](processor.asValueMapperWithKey)
+  }
+
+  def print(printed: Printed[K, V]): Unit = inner.print(printed)
+
+  def foreach(action: (K, V) => Unit): Unit = {
+    inner.foreach((k, v) => action(k, v))
+  }
+
+  def branch(predicates: ((K, V) => Boolean)*): Array[KStream[K, V]] = {
+    inner.branch(predicates.map(_.asPredicate): _*).map(kstream => wrapKStream(kstream))
+  }
+
+  def through(topic: String)(implicit produced: Produced[K, V]): KStream[K, V] =
+    inner.through(topic, produced)
+
+  def to(topic: String)(implicit produced: Produced[K, V]): Unit =
+    inner.to(topic, produced)
+
+  //scalastyle:off null
+  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, (K1, V1)],
+    stateStoreNames: String*): KStream[K1, V1] = {
+
+    val transformerSupplierJ: TransformerSupplier[K, V, KeyValue[K1, V1]] = () => {
+      val transformerS: Transformer[K, V, (K1, V1)] = transformerSupplier()
+      new Transformer[K, V, KeyValue[K1, V1]] {
+        override def transform(key: K, value: V): KeyValue[K1, V1] = {
+          transformerS.transform(key, value) match {
+            case (k1, v1) => KeyValue.pair(k1, v1)
+            case _ => null
+          }
+        }
+
+        override def init(context: ProcessorContext): Unit = transformerS.init(context)
+
+        @deprecated ("Please use Punctuator functional interface at https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/processor/Punctuator.html instead", "0.1.3") // scalastyle:ignore
+        override def punctuate(timestamp: Long): KeyValue[K1, V1] = {
+          transformerS.punctuate(timestamp) match {
+            case (k1, v1) => KeyValue.pair[K1, V1](k1, v1)
+            case _ => null
+          }
+        }
+
+        override def close(): Unit = transformerS.close()
+      }
+    }
+    inner.transform(transformerSupplierJ, stateStoreNames: _*)
+  }
+  //scalastyle:on null
+
+  def transformValues[VR](valueTransformerSupplier: () => ValueTransformer[V, VR],
+    stateStoreNames: String*): KStream[K, VR] = {
+
+    val valueTransformerSupplierJ: ValueTransformerSupplier[V, VR] = () => valueTransformerSupplier()
```

Comment:
A `ValueTransformer` also have `init()`, `punctuate()` and `close()` method. Why is this code much simpler than the wrapper for `transform()` above?

Reply:
We don't need to provide implementation of `ValueTransformer` here since the passed in `() => ValueTransformer[V, VR]` gets converted to `ValueTransformerSupplier[V, VR]` in the implementation through SAM type conversion. We could not do that for `transform` as the SAM type conversion will not be handled automatically in that case.